File: oops/matrixmultiplication.cpp
#include <iostream>
using namespace std;

class Matrix
{
private:
    int **matrix;
    int rows;
    int cols;

public:
    // Constructor to initialize matrix
    Matrix(int r, int c)
    {
        rows = r;
        cols = c;
        matrix = new int *[rows];
        for (int i = 0; i < rows; i++)
        {
            matrix[i] = new int[cols];
        }
    }

    // Destructor to free memory
    ~Matrix()
    {
        for (int i = 0; i < rows; i++)
        {
            delete[] matrix[i];
        }
        delete[] matrix;
    }

    // Function to input values for matrix
    void getData()
    {
        cout << "Enter elements of the matrix: " << endl;
        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                cin >> matrix[i][j];
            }
        }
    }

    // Function to display the matrix
    void show()
    {
        cout << "Matrix: " << endl;
        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                cout << matrix[i][j] << " ";
            }
            cout << endl;
        }
    }

    // Function to multiply two matrices
    Matrix *mul(Matrix *m)
    {
        if (cols != m->rows)
        {
            cout << "Matrices cannot be multiplied." << endl;
            return NULL;
        }

        Matrix *result = new Matrix(rows, m->cols);
        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < m->cols; j++)
            {
                result->matrix[i][j] = 0;
                for (int k = 0; k < cols; k++)
                {
                    result->matrix[i][j] += matrix[i][k] * m->matrix[k][j];
                }
            }
        }
        return result;
    }
};

int main()
{
    int r1, c1, r2, c2;
    cout << "Enter number of rows and columns for first matrix: ";
    cin >> r1 >> c1;
    cout << "Enter number of rows and columns for second matrix: ";
    cin >> r2 >> c2;

    Matrix *m1 = new Matrix(r1, c1);
    Matrix *m2 = new Matrix(r2, c2);

    m1->getData();
    m2->getData();

    cout << "First Matrix: " << endl;
    m1->show();
    cout << "Second Matrix: " << endl;
    m2->show();

    Matrix *result = m1->mul(m2);
    if (result != NULL)
    {
        cout << "Resultant Matrix: " << endl;
        result->show();
        delete result;
    }

    delete m1;
    delete m2;

    return 0;
}

File: java/java codes/MatrixMultiplication.java
import java.util.Scanner;

public class MatrixMultiplication {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Input for the first matrix
        System.out.print("Enter the number of rows for matrix 1: ");
        int m = sc.nextInt();
        System.out.print("Enter the number of columns for matrix 1: ");
        int n = sc.nextInt();

        int[][] a = new int[m][n];

        System.out.println("Enter the elements for matrix 1:");
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                a[i][j] = sc.nextInt();
            }
        }

        // Input for the second matrix
        System.out.print("Enter the number of rows for matrix 2: ");
        int o = sc.nextInt();
        System.out.print("Enter the number of columns for matrix 2: ");
        int p = sc.nextInt();

        int[][] b = new int[o][p];

        System.out.println("Enter the elements for matrix 2:");
        for (int i = 0; i < o; i++) {
            for (int j = 0; j < p; j++) {
                b[i][j] = sc.nextInt();
            }
        }

        // Matrix multiplication condition
        if (n == o) {
            // Matrix multiplication
            int[][] result = new int[m][p];
            
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < p; j++) {
                    result[i][j] = 0;
                    for (int k = 0; k < n; k++) {  // (n== o) 
                        result[i][j] += a[i][k] * b[k][j];
                    }
                }
            }

            // Display the result
            System.out.println("\nResultant matrix after multiplication:");
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < p; j++) {
                    System.out.print(result[i][j] + " ");
                }
                System.out.println();
            }
        } else {
            System.out.println("Matrix cannot be multiplied!");
        }

        sc.close();
    }
}
